go/tools/asthelpergen/integration/types.go

```diff
@@ -0,0 +1,172 @@
+To recreate them, just run:
+
+go run go/tools/asthelpergen -in ./go/tools/asthelpergen/integration -iface vitess.io/vitess/go/tools/asthelpergen/integration.AST
+*/
+// AST is the interface all interface types implement
+type AST interface {
+	String() string
+}
+
+// Empty struct impl of the iface
+type Leaf struct {
+	v int
+}
+
+func (l *Leaf) String() string {
+	if l == nil {
+		return "nil"
+	}
+	return fmt.Sprintf("Leaf(%d)", l.v)
+}
+
+// Container implements the interface ByRef
+type RefContainer struct {
+	ASTType               AST
+	NotASTType            int
+	ASTImplementationType *Leaf
+}
+
+func (r *RefContainer) String() string {
+	if r == nil {
+		return "nil"
+	}
+	var astType = ""
+	if r.ASTType == nil {
+		astType = "nil"
+	} else {
+		astType = r.ASTType.String()
+	}
+	return fmt.Sprintf("RefContainer{%s, %d, %s}", astType, r.NotASTType, r.ASTImplementationType.String())
+}
+
+// Container implements the interface ByRef
+type RefSliceContainer struct {
+	ASTElements               []AST
+	NotASTElements            []int
+	ASTImplementationElements []*Leaf
+}
+
+func (r *RefSliceContainer) String() string {
+	return fmt.Sprintf("RefSliceContainer{%s, %s, %s}", sliceStringAST(r.ASTElements...), "r.NotASTType", sliceStringLeaf(r.ASTImplementationElements...))
+}
+
+// Container implements the interface ByValue
+type ValueContainer struct {
+	ASTType               AST
+	NotASTType            int
+	ASTImplementationType *Leaf
+}
+
+func (r ValueContainer) String() string {
+	return fmt.Sprintf("ValueContainer{%s, %d, %s}", r.ASTType.String(), r.NotASTType, r.ASTImplementationType.String())
+}
+
+// Container implements the interface ByValue
+type ValueSliceContainer struct {
+	ASTElements               []AST
+	NotASTElements            []int
+	ASTImplementationElements []*Leaf
+}
+
+func (r ValueSliceContainer) String() string {
+	return fmt.Sprintf("ValueSliceContainer{%s, %s, %s}", sliceStringAST(r.ASTElements...), "r.NotASTType", sliceStringLeaf(r.ASTImplementationElements...))
+}
+
+// We need to support these types - a slice of AST elements can implement the interface
+type InterfaceSlice []AST
+
+func (r InterfaceSlice) String() string {
+	var elements []string
+	for _, el := range r {
+		elements = append(elements, el.String())
+	}
+
+	return "[" + strings.Join(elements, ", ") + "]"
+}
+
+// We need to support these types - a slice of AST elements can implement the interface
+type Bytes []byte
+
+func (r Bytes) String() string {
+	return string(r)
+}
+
+type LeafSlice []*Leaf
+
+func (r LeafSlice) String() string {
+	var elements []string
+	for _, el := range r {
+		elements = append(elements, el.String())
+	}
+	return strings.Join(elements, ", ")
+}
+
+type BasicType int
+
+func (r BasicType) String() string {
+	return fmt.Sprintf("int(%d)", r)
+}
+
+const (
+	// these consts are here to try to trick the generator
+	thisIsNotAType  BasicType = 1
+	thisIsNotAType2 BasicType = 2
+)
```

Comment:
the comment is not lying. these consts produce code that we had to do extra checks in the generator code to avoid. They are here to mimic sqlparser types such as `ReadOnly` (https://github.com/vitessio/vitess/blob/master/go/vt/sqlparser/constants.go#L254)